### ISSUE TITLE
Low level functions ScriptToAsmStr (core_io), i64tostr (strencodings) and itostr (strencodings) are locale dependent

### ISSUE THREAD

**Issue by practicalswift** [bitcoin/bitcoin#18281]
Low level functions `ScriptToAsmStr` (`core_io`), `i64tostr` (`strencodings`) and `itostr` (`strencodings`) are locale dependent:

```c++
const std::vector<uint8_t> b{0x6a, 0x4, 0xff, 0xff, 0xff, 0xff};
const CScript script{b.begin(), b.end()};
for (const std::string& locale_string : {"C", "de_DE"}) {
    std::locale::global(std::locale(locale_string));
    std::cout << "[" << locale_string << "] ScriptToAsmStr(script, false) == " 
        << ScriptToAsmStr(script, false) << "\n";
    std::cout << "[" << locale_string << "] i64tostr(12345678) == " 
        << i64tostr(12345678) << "\n";
    std::cout << "[" << locale_string << "] itostr(12345678) == " 
        << itostr(12345678) << "\n";
}
```

```
[C] ScriptToAsmStr(script, false) == OP_RETURN -2147483647
[C] i64tostr(12345678) == 12345678
[C] itostr(12345678) == 12345678
[de_DE] ScriptToAsmStr(script, false) == OP_RETURN -2.147.483.647
[de_DE] i64tostr(12345678) == 12.345.678
[de_DE] itostr(12345678) == 12.345.678
```

**Comment by MarcoFalke** [comment]
I can't reproduce with bitcoind or bitcoin-qt. 


![Screenshot from 2020-03-06 14-40-21](https://user-images.githubusercontent.com/6399679/76116684-70d74a00-5fb8-11ea-9fbb-fa692e96b66f.png)

**Comment by practicalswift** [comment]
@MarcoFalke That's because the _C++_ locale is _currently_ not reflecting `LC_*` which appears to be by pure luck and not thanks to a conscious decision: [see the comments in the PR #18124 ("init: Clarify C and C++ locale assumptions. Add locale sanity checks to verify assumptions at run-time.") for context](https://github.com/bitcoin/bitcoin/pull/18124). Please read the entire thread to get all the gotchas: there are quite a few of them :)

Note that in `bitcoin-qt` the C locale (`setlocale`) is reflecting `LC_*`, but the C++ locale (`std::locale::global`) is not.

If we consider `ScriptToAsmStr`, `i64tostr` and `itostr` to be safe as written now that implies that we assume that the C++ locale never being set to anything but `C` (the classic locale).

If we assume the C++ locale to always be `C` then we should state that explicitly and also assert it as runtime as suggested in #18124.

The current non-principled "whack-a-mole" approach to locale handling seems fragile at best :)

**Comment by practicalswift** [comment]
This one could be marked "good first issue": it should be trivial to fix.

**Comment by MarcoFalke** [comment]
@practicalswift "good first issue" is only for issues where one (and only one) uncontroversial solution exists.

**Comment by MarcoFalke** [comment]
Also `i64tostr` doesn't exist in this repo. I think the issue has already been solved?

**Comment by practicalswift** [comment]
@MarcoFalke Yes, `i64tostr` and `itostr` are gone, but `ScriptToAsmStr` is still doing `strprintf("%d", CScriptNum(vch, false).getint());` which is locale dependent:

```c++
const std::vector<uint8_t> b{0x6a, 0x4, 0xff, 0xff, 0xff, 0xff};
const CScript script{b.begin(), b.end()};
for (const std::string& locale_string : {"C", "de_DE"}) {
    std::locale::global(std::locale(locale_string));
    std::cout << "[" << locale_string << "] ScriptToAsmStr(script, false) == " 
        << ScriptToAsmStr(script, false) << "\n";
}
```

```
[C] ScriptToAsmStr(script, false) == OP_RETURN -2147483647
[de_DE] ScriptToAsmStr(script, false) == OP_RETURN -2.147.483.647
```

:)

**Comment by MarcoFalke** [comment]
yeah, and that is fixed by #18450
